util/topsql/reporter/reporter.go

```diff
@@ -0,0 +1,411 @@
+	"google.golang.org/grpc/backoff"
+)
+
+// TopSQLReporter collects Top SQL metrics.
+type TopSQLReporter interface {
+	tracecpu.Reporter
+	RegisterSQL(sqlDigest []byte, normalizedSQL string)
+	RegisterPlan(planDigest []byte, normalizedPlan string)
+}
+
+type topSQLCPUTimeInput struct {
+	timestamp uint64
+	records   []tracecpu.TopSQLCPUTimeRecord
+}
+
+type planBinaryDecodeFunc func(string) (string, error)
+
+// topSQLDataPoints represents the cumulative SQL plan CPU time in current minute window
+type topSQLDataPoints struct {
+	SQLDigest      []byte
+	PlanDigest     []byte
+	CPUTimeMsList  []uint32
+	TimestampList  []uint64
+	CPUTimeMsTotal uint64
+}
+
+// cpuTimeSort is used to sort TopSQL records by tocal CPU time
+type cpuTimeSort struct {
+	Key            string
+	SQLDigest      []byte
+	PlanDigest     []byte
+	CPUTimeMsTotal uint64
+}
+type cpuTimeSortSlice []cpuTimeSort
+
+func (t cpuTimeSortSlice) Len() int {
+	return len(t)
+}
+
+// We need find the kth largest value, so here should use >
+func (t cpuTimeSortSlice) Less(i, j int) bool {
+	return t[i].CPUTimeMsTotal > t[j].CPUTimeMsTotal
+}
+func (t cpuTimeSortSlice) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+type planRegisterJob struct {
+	planDigest     string
+	normalizedPlan string
+}
+
+// TopSQLReporterImpl is called periodically to collect TopSQL resource usage metrics
+type TopSQLReporterImpl struct {
+	mu sync.RWMutex
+	// calling this can take a while, so should not block critical paths
+	planBinaryDecoder planBinaryDecodeFunc
+
+	// topSQLMap maps `sqlDigest-planDigest` to TopSQLDataPoints
+	topSQLMap        map[string]*topSQLDataPoints
+	MaxStatementsNum int
+
+	// normalizedSQLMap is an map, whose keys are SQL digest strings and values are normalized SQL strings
+	normalizedSQLMap map[string]string
+
+	// normalizedPlanMap is a plan version of normalizedSQLMap
+	// this should only be set from the dedicated worker
+	normalizedPlanMap map[string]string
+
+	collectCPUTimeChan chan *topSQLCPUTimeInput
+	planRegisterChan   chan *planRegisterJob
+
+	collectInterval  time.Duration
+	collectTimeout   time.Duration
+	agentGRPCAddress string
+
+	// current tidb-server instance ID
+	instanceID string
+
+	quit chan struct{}
+}
+
+// TopSQLReporterConfig is the config for TopSQLReporterImpl
+type TopSQLReporterConfig struct {
+	// PlanBinaryDecoder is used to decode the plan binary before sending to agent
+	PlanBinaryDecoder planBinaryDecodeFunc
+	// MaxStatementsNum is the capacity of the TopSQL map, above which evition should happen
+	MaxStatementsNum int
+	// CollectInterval is the interval of sending TopSQL records to the agent
+	CollectInterval time.Duration
+	// CollectTimeout is the timeout of a single agent gRPC call
+	CollectTimeout time.Duration
+	// AgentGRPCAddress is the gRPC address of the agent server
+	AgentGRPCAddress string
+	// InstanceID is the ID of this tidb server
+	InstanceID string
+}
+
+func encodeCacheKey(sqlDigest, planDigest []byte) []byte {
```

Comment:
- You can change the name, as it is no longer a cache. Maybe just `encodeKey`.
- We can allocate the buffer with a fixed size beforehand. Otherwise this buffer may be allocated multiple times to grow.
- We can grab the `Buffer` from a `sync.Pool` and throw out a `string`. In this way the sync.Pool small objects will be reused very efficiently

Reply:
How about return `string` instead of `[]byte`, to avoid multi `string(key)` in other places.